apps/studio/components/ui/QueryBlock/QueryBlock.tsx

```diff
@@ -0,0 +1,338 @@
+            tooltip={{
+              content: {
+                side: 'bottom',
+                className: 'max-w-56 text-center',
+                text: isExecuting ? (
+                  <p>{`Query is running. You may cancel ongoing queries via the [SQL Editor](/project/${ref}/sql?viewOngoingQueries=true).`}</p>
+                ) : (
+                  'Run query'
+                ),
+              },
+            }}
+          />
+
+          {/* [Joshen ReportsV2] Un-tested as this is intended for Reports V2 */}
+          {!disableUpdate && onRemoveChart && (
+            <ButtonTooltip
+              type="text"
+              size="tiny"
+              icon={<X />}
+              className="w-7 h-7"
+              onClick={() => onRemoveChart({ metric: { key: id ?? '' } })}
+              tooltip={{ content: { side: 'bottom', text: 'Remove chart' } }}
+            />
+          )}
+        </div>
+      </div>
+
+      {showWarning && (
+        <Admonition
+          type="warning"
+          className="mb-0 rounded-none border-0 shrink-0 bg-background-100 border-t"
+        >
+          <p>
+            {showWarning === 'hasWriteOperation'
+              ? 'This query contains write operations.'
+              : 'This query involves running a function.'}{' '}
+            Are you sure you want to execute it?
+          </p>
+          <p className="text-foreground-light">
+            Make sure you are not accidentally removing something important.
+          </p>
+          <div className="flex justify-stretch mt-2 gap-2">
+            <Button
+              type="outline"
+              size="tiny"
+              className="w-full flex-1"
+              onClick={() => setShowWarning(undefined)}
+            >
+              Cancel
+            </Button>
+            <Button
+              type="danger"
+              size="tiny"
+              className="w-full flex-1"
+              onClick={() => {
+                setShowWarning(undefined)
+                const processedSql = processParameterizedSql(sql!, combinedParameterValues)
+                execute({
+                  projectRef: ref,
+                  connectionString: project?.connectionString,
+                  sql: processedSql,
+                })
+              }}
+            >
+              Run
+            </Button>
+          </div>
+        </Admonition>
+      )}
+
+      {/* QueryBlock output */}
+      {showSql && (
+        <div className="shrink-0 max-h-96 overflow-y-auto border-t">
+          <CodeBlock
+            hideLineNumbers
+            wrapLines={false}
+            value={sql}
+            language="sql"
+            className={cn(
+              'max-w-none block !bg-transparent !py-3 !px-3.5 prose dark:prose-dark border-0 border-t text-foreground !rounded-none w-full',
+              '[&>code]:m-0 [&>code>span]:text-foreground'
+            )}
+          />
+        </div>
+      )}
+
+      {queryResult === undefined && (
```

Comment:
For queries that aren't run on load, like those that have runQuery set to false or mutable queries, this will always show.

```suggestion
      {isExecuting && !queryResult && (
```

Reply:
good catch! 🙏